proto/src/rr/domain/name.rs

```diff
@@ -117,8 +109,8 @@ impl Name {
     /// let name = name.append_label("com");
     /// assert_eq!(name, Name::from_str("www.example.com").unwrap());
     /// ```
-    pub fn append_label<S: Into<String>>(mut self, label: S) -> Self {
-        self.labels.push(Rc::new(label.into()));
+    pub fn append_label<S: AsRef<str>>(mut self, label: S) -> Self {
```

Comment:
This should be named `append_idna_label` or something more specific. In particular, it should always be explicit and obvious what the expected encoding is.

Reply:
This is good feedback, though I'm not sure of the desired outcome is that you would like. Right now the logic is to first check if it's ascii, then encode to IDNA if it is not. Basically the outcome is that if it's plain ascii it will work as expected, if it is not then it will become IDNA.

What I really need to do is change these interfaces to take `Into<Label>` and then implement that for `&[u8]`, `String`, and `&str`. That would allow for more control over the type of label being introduced.

Reply:
I think people may prefer two functions with differing semantics:

1. Encode UTF-8 into IDNA.
2. Require that the input is already IDNA and fail if it isn't.

In particular, in many/most cases the caller will be using a string that they expect to be IDNA and they may be expecting that the label will be exactly the same as their input string; i.e. a conversion would be surprising and potentially cause them problems.

Anyway, probably a better way to do this would be to have append_label take a `Label` as input, and then the various conversions can be done only in `Label`.

Reply:
I'll change all `Name::*label*` methods to take `Into<Label>`.

> I think people may prefer two functions with differing semantics:

I'll clean up the construction of Label for these two cases: 

- `from_utf8(AsRef<str>)`, which will always perform idna conversion (We should probably migrate `Name::from_str` to this as well). Will fail on bad conversion. 
- `from_ascii(AsRef<[u8]>)`, which will fail if any portion of the label is outside DNS acceptable ascii.
- `from_raw_bytes(&[u8])`, create an arbitrary binary label (not generally useful outside serialization). This shouldn't be outside standard ascii of `from_ascii`, but we can get anything off the wire...